sys/psa_crypto/psa_persistent_storage/psa_crypto_cbor_encoder.c

```diff
@@ -0,0 +1,625 @@
+        if (res < 0) {
+            goto error;
+        }
+    }
+
+    return PSA_SUCCESS;
+
+error:
+    DEBUG("[persistent storage] error while encoding key data: %d = %s, ", res,
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+
+#if PSA_PROTECTED_KEY_COUNT
+/**
+ * @brief   Encode contents of a PSA key slot containing a protected key
+ *
+ * @param   enc     Active nanocbor encoder
+ * @param   slot    Pointer to key slot to encode
+ * @return  psa_status_t
+ */
+static psa_status_t psa_encode_protected_key_slot(nanocbor_encoder_t *enc, psa_key_slot_t *slot)
+{
+    int res;
+    int key_array_size = 1;
+    psa_key_slot_number_t *slot_no = psa_key_slot_get_slot_number(slot);
+
+#if IS_USED(MODULE_PSA_ASYMMETRIC)
+    if (PSA_KEY_TYPE_IS_KEY_PAIR(slot->attr.type)) {
+        key_array_size = 2;
+    }
+#endif /* MODULE_PSA_ASYMMETRIC */
+
+    res = nanocbor_fmt_array(enc, key_array_size);
+    if (res < 0) {
+        goto error;
+    }
+
+    res = nanocbor_fmt_uint(enc, *slot_no);
+    if (res < 0) {
+        goto error;
+    }
+
+#if IS_USED(MODULE_PSA_ASYMMETRIC)
+    if (PSA_KEY_TYPE_IS_KEY_PAIR(slot->attr.type)) {
+        uint8_t *pubkey_data;
+        size_t *pubkey_data_len;
+        psa_get_public_key_data_from_key_slot(slot, &pubkey_data, &pubkey_data_len);
+
+        res = nanocbor_fmt_array(enc, 2);
+        if (res < 0) {
+            goto error;
+        }
+
+        psa_status_t status = psa_encode_key_data(enc, pubkey_data, *pubkey_data_len);
+        if (status != PSA_SUCCESS) {
+            return status;
+        }
+    }
+#endif /* MODULE_PSA_ASYMMETRIC */
+
+    return PSA_SUCCESS;
+
+error:
+    DEBUG("[persistent storage] error while encoding protected key: %d = %s", res,
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+#endif /* PSA_PROTECTED_KEY_COUNT */
+
+#if PSA_ASYMMETRIC_KEYPAIR_COUNT
+/**
+ * @brief   Encode contents of a PSA key slot containing an asymmetric key pair
+ *
+ * @param   enc     Active nanocbor encoder
+ * @param   slot    Pointer to key slot to encode
+ * @return  psa_status_t
+ */
+static psa_status_t psa_encode_asymmetric_key_pair(nanocbor_encoder_t *enc, psa_key_slot_t *slot)
+{
+    int res;
+    psa_status_t status;
+    int key_array_size = 2; /* Contains public and private key */
+    size_t *privkey_data_len;
+    size_t *pubkey_data_len;
+    uint8_t *privkey_data;
+    uint8_t *pubkey_data;
+
+    psa_get_key_data_from_key_slot(slot, &privkey_data, &privkey_data_len);
+    psa_get_public_key_data_from_key_slot(slot, &pubkey_data, &pubkey_data_len);
+
+    /* Format Key Pair Array */
+    res = nanocbor_fmt_array(enc, key_array_size);
+    if (res < 0) {
+        goto error;
+    }
+
+    /* Format Private Key Array containing key size and key data */
+    res = nanocbor_fmt_array(enc, key_array_size);
```

Comment:
This should probably be a hard-coded 2 (or better moved inside `psa_encode_key_data`) as argued above.